### ISSUE TITLE
Cluster by arbitrary shapes (NUTS1, NUTS2, NUTS3, GADM)

### ISSUE THREAD

**Issue by fneum** [PyPSA/pypsa-eur#263]
It is already possible to specify a custom busmap to the `cluster_network` rule to support arbitrary clustering. But that is so far a manual process.

There are many use cases for clustering according to administrative areas (NUTS, GADM, etc).

These are usually available as shapefiles (e.g. https://gisco-services.ec.europa.eu/distribution/v2/nuts/download/#nuts21).

I would like to automate clustering to arbitrary shapefiles. That could look something like this:

To create (note that you need not to supply a number of clusters):

```sh
snakemake -j all "networks/elec_s_shapes.nc"
```

Provide a shapefile in `config.yaml`:

```yaml
clustering:
  optional_inputs:
    shapefile: "nuts/NUTS_RG_01M_2016_4326_LEVL_0.geojson"
```

That becomes and optional input to the `snakemake` rule in the `Snakefile`:

```smk
wildcard_constraints:
    clusters="[0-9]+m?|all|shapes",

rule cluster_network:
    input:
        **config["clustering"].get("optional_inputs", {})
```

I think the busmap should be relatively straightforward to create using `geopandas.sjoin()`. One would add a function to `scripts/cluster_network.py`:

```py
def busmap_from_shapes():
    ...
```

Two difficulties I can think of right now:

- The mapping of a bus' Voronoi cell from the base network to the regions datafile is not always 100% clear; i.e. the borders of regions will run through Voronoi cells which need to be split/disaggregated. Vice versa, it can also occur that a region is fully contained inside a Voronoi cell of the base network (presumably less likely).
- Offshore nodes need to be mapped to onshore buses (that are covered by shapefile)

**Comment by fneum** [comment]
Example for NUTS1 and region shapes from `elec_s.nc`:
![image](https://user-images.githubusercontent.com/29101152/125187986-886cb580-e232-11eb-9c2b-15096fcd73d0.png)

**Comment by pz-max** [comment]
Hi Fabian, the `build_bus_region.py` creates the Voronoi cells to which all information are aggregated.
Maybe it is worth adjusting the `build_bus_region.py` instead of the `cluster_network.py` function? 

Or is there a particular reason for adjusting `cluster_network.py`?

**Comment by FabianHofmann** [comment]
But the Voronoi cells are build on basis of the ENTSOE gridmap, i.e. their nodes. I think the problem remains the same, as the nodes of the ENTOSE maps have to be aligned to the NUTS3 regions...

**Comment by pz-max** [comment]
Yes, true. The problem stays the same. What was nice with the Voronoi bus map technique is that each region had a clear connection. One node was connected to another one with a line. The main question is now with NUTS/GADM or other arbitrary shapes. How are these regions connected? Which region is linked to each with which magnitude?

In general what can happen:
- A region is isolated (only transmission/distribution infrastructure in the region itself)
- A region is connected with other regions

We would do the following:
1) Create ID's for any shape i.e. Voronoi_busmap/GADM/NUTS.
2) If buses/lines are in the same shape -> Associate the capacities and cost at the region as i.e. "distribution"
3) If two buses, connected by a line, are in two different regions. Define a connection between the region with the given transmission capacity and costs.

One could then also develop some nice search algorithms i.e.
- Connect from different locations to the isolated regions and assess if it reduces the total system costs (also DC lines connected from far distant regions)
Or a non-search approach/option:
- Define all regions as connected to the neighbouring ones (This is a good default which I guess is applied in Calliope)

**Comment by fneum** [comment]
Hmmm, I don't really see a good way of defining the clustered regions further upstream without a tremendous amount of work. Everything is tied around the ENTSO-E grid map.

- The network simplification (aligning voltage level, removing stubs etc) in `simplify_network` would have to be moved further upstream, too.
- If the initial regions become too large `build_renewable_profiles` will likely get a performance hit because the rasters need to be chunked for calculating the renewable potentials.

**Comment by FabianHofmann** [comment]
> * If the initial regions become too large `build_renewable_profiles` will likely get a performance hit because the rasters need to be chunked for calculating the renewable potentials.

This should not be a problem, the performance with large regions is pretty solid. But anyhow, it would require a lot of rearrangement considering the simplify_network and correct assignment of load time series.

**Comment by fneum** [comment]
@Hazem-IEG, were you planning to transfer this also to PyPSA-Eur at some point?

**Comment by Hazem-IEG** [comment]
@fneum, yes, I already started working on it locally and did some initial tests a while back but stopped to focus on the sector coupled version of PyPSA-Earth.. soon I'll continue working on it

**Comment by fneum** [comment]
@Hazem-IEG any updates? This would be a nice feature to add to PyPSA-Eur.

**Comment by fneum** [comment]
This is in PyPSA-Earth:

Look for "alternative clustering":

https://github.com/pypsa-meets-earth/pypsa-earth/blob/dbca213b43bf96fa25ec5b4785cddadec3a85f5c/scripts/build_bus_regions.py#L200-L207

https://github.com/pypsa-meets-earth/pypsa-earth/blob/main/scripts/cluster_network.py

https://github.com/pypsa-meets-earth/pypsa-earth/blob/main/scripts/simplify_network.py

**Comment by fneum** [comment]
Also https://www.geoboundaries.org/globalDownloads.html

**Comment by fneum** [comment]
Also @bobbyxng worked this out in PyPSA-Eur: https://github.com/bobbyxng/kodekiste/tree/main/scripts/cluster-by-nuts

**Comment by pworschischek-aggmag** [comment]
Hi there @bobbyxng and @fneum 👋 

I wanted to check in on this feature and ask if you need help with it. I met @lindnemi at IEWT last week, which was nice and he encouraged me to simply write to you here. Our use case would hugely benefit from the shape clustering, so I already ran the current draft locally. However, I find a few things still need mending. Let me know if you can think of ways I can be of help. Tests, docs, identifying edge cases, something along those lines, you name it.

**Comment by fneum** [comment]
Yes, it's currently underway here: https://github.com/PyPSA/pypsa-eur/pull/1502

**Comment by pworschischek-aggmag** [comment]
That's great news, let me know if I can do something to help you out.

**Comment by bobbyxng** [comment]
> That's great news, let me know if I can do something to help you out.

Thanks for your interest! The implementation/PR should be ready now and is under review. You can also check-out the PR and do your own tests, if you feel like doing so :)

To activate clustering by administrative levels/NUTS, set the clusters wildcard to "adm"

```
scenario:
  clusters:
  - adm
```
and
```
clustering:
  mode: administrative
  administrative:
    level: 1 # levels 0-3
    # DE: 2 # optional
    # FR: 2 # optional
```
